Source/WebCore/html/FormAssociatedCustomElement.cpp

```diff
@@ -0,0 +1,279 @@
+
+    if (!m_submissionValue.has_value())
+        return false;
+
+    WTF::switchOn(m_submissionValue.value(), [&](RefPtr<DOMFormData> value) {
+        for (const auto& item : value->items()) {
+            WTF::switchOn(item.data, [&](const String& value) {
+                formData.append(item.name, value);
+            }, [&](RefPtr<File> value) {
+                formData.append(item.name, *value);
+            });
+        }
+    }, [&](const String& value) {
+        if (!name().isEmpty())
+            formData.append(name(), value);
+    }, [&](RefPtr<File> value) {
+        if (!name().isEmpty())
+            formData.append(name(), *value);
+    });
+
+    return true;
+}
+
+void FormAssociatedCustomElement::formWillBeDestroyed()
+{
+    m_formWillBeDestroyed = true;
+    ValidatedFormListedElement::formWillBeDestroyed();
+    m_formWillBeDestroyed = false;
+}
+
+bool FormAssociatedCustomElement::isEnumeratable() const
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    return true;
+}
+
+void FormAssociatedCustomElement::reset()
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    CustomElementReactionQueue::enqueueFormResetCallbackIfNeeded(*m_element);
+}
+
+void FormAssociatedCustomElement::disabledStateChanged()
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    ValidatedFormListedElement::disabledStateChanged();
+    CustomElementReactionQueue::enqueueFormDisabledCallbackIfNeeded(*m_element, isDisabled());
+}
+
+void FormAssociatedCustomElement::didChangeForm()
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    ValidatedFormListedElement::didChangeForm();
+    if (!m_formWillBeDestroyed)
+        CustomElementReactionQueue::enqueueFormAssociatedCallbackIfNeeded(*m_element, form());
+}
+
+void FormAssociatedCustomElement::willUpgrade()
+{
+    setDataListAncestorState(TriState::False);
+}
+
+void FormAssociatedCustomElement::didUpgrade()
+{
+    ASSERT(!form());
+
+    HTMLElement& element = asHTMLElement();
+
+    parseFormAttribute(element.attributeWithoutSynchronization(formAttr));
+    parseDisabledAttribute(element.attributeWithoutSynchronization(disabledAttr));
+    parseReadonlyAttribute(element.attributeWithoutSynchronization(readonlyAttr));
+
+    setDataListAncestorState(TriState::Indeterminate);
+    updateWillValidateAndValidity();
+    syncWithFieldsetAncestors(element.parentNode());
+    invalidateElementsCollectionCachesInAncestors();
+    restoreFormControlStateIfNecessary();
+}
+
+void FormAssociatedCustomElement::finishParsingChildren()
+{
+    if (!FormController::ownerForm(*this))
+        restoreFormControlStateIfNecessary();
+}
+
+void FormAssociatedCustomElement::invalidateElementsCollectionCachesInAncestors()
+{
+    auto invalidateElementsCache = [](HTMLElement& element) {
+        if (auto* nodeLists = element.nodeLists())
+            nodeLists->invalidateCachesForAttribute(HTMLNames::formAttr);
+    };
+
+    if (RefPtr<HTMLFormElement> form = this->form())
```

Comment:
Nit: we can just say RefPtr. No need to spell out HTMLFormElement.

Reply:
Simplified here and in a few other places.